src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs

```diff
@@ -8,31 +8,33 @@
 using System.Linq;
 using NuGet.Shared;
 
+#nullable enable
+
 namespace NuGet.RuntimeModel
 {
-    public class RuntimeGraph : IEquatable<RuntimeGraph>
+    public sealed class RuntimeGraph : IEquatable<RuntimeGraph>
     {
-        private readonly ConcurrentDictionary<RuntimeCompatKey, bool> _areCompatible
-            = new ConcurrentDictionary<RuntimeCompatKey, bool>();
-
-        private readonly ConcurrentDictionary<string, HashSet<string>> _expandCache
-            = new ConcurrentDictionary<string, HashSet<string>>(StringComparer.Ordinal);
-
-        private readonly ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>> _dependencyCache
-            = new ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>>();
+        // These fields are null when IsEmpty is true
+        private readonly ConcurrentDictionary<RuntimeCompatKey, bool>? _areCompatible;
+        private readonly ConcurrentDictionary<string, HashSet<string>>? _expandCache;
+        private readonly ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>>? _dependencyCache;
 
-        private HashSet<string> _packagesWithDependencies;
+        private HashSet<string>? _packagesWithDependencies;
 
         public static readonly string RuntimeGraphFileName = "runtime.json";
 
-        public static readonly RuntimeGraph Empty = new RuntimeGraph();
+        /// <summary>
+        /// Gets a singleton, immutable, empty instance of <see cref="RuntimeGraph"/>.
+        /// </summary>
+        public static readonly RuntimeGraph Empty = new();
 
         public IReadOnlyDictionary<string, RuntimeDescription> Runtimes { get; }
-        public IReadOnlyDictionary<string, CompatibilityProfile> Supports { get; set; }
+        public IReadOnlyDictionary<string, CompatibilityProfile> Supports { get; }
```

Comment:
This setter was removed. Note that `Runtimes` didn't have one to remove. These objects are otherwise immutable, and the optimisations here require that immutability. There were no usages of this setter in this repo.

Reply:
Unclear to me why https://github.com/NuGet/NuGet.Client/commit/3dc89fbb2b29ca8f05b797d8d49440931e45b348 added a setter. It wasn't used in that PR either.